slam-libraries/viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -420,7 +432,74 @@ std::string SLAMServiceImpl::GetNextDataFile() {
     return GetNextDataFileOnline();
 }
 
-void SLAMServiceImpl::RunSLAM() {
+void SLAMServiceImpl::StartSaveMap() {
+    if (map_rate_sec == std::chrono::seconds(0)) {
+        return;
+    }
+    thread_save_map_with_timestamp =
+        new std::thread([&]() { this->SaveMapWithTimestamp(); });
+}
+
+void SLAMServiceImpl::StopSaveMap() {
+    if (map_rate_sec == std::chrono::seconds(0)) {
+        return;
+    }
+    thread_save_map_with_timestamp->join();
+}
+
+void SLAMServiceImpl::SaveMapWithTimestamp() {
+    auto check_for_shutdown_interval_usec =
+        std::chrono::microseconds(checkForShutdownIntervalMicroseconds);
+    while (b_continue_session) {
+        auto start = std::chrono::high_resolution_clock::now();
+        const std::string filename_with_timestamp =
+            viam::io::MakeFilenameWithTimestamp(path_to_map);
+
+        if (offlineFlag && finished_processing_offline) {
+            {
+                std::lock_guard<std::mutex> lk(map_builder_mutex);
+                // Save the map in a pbstream file
+                bool ok = map_builder.map_builder_->SerializeStateToFile(
+                    true, filename_with_timestamp);
+                if (!ok) {
+                    LOG(WARNING) << "Saving the map to pbstream failed.";
+                }
+            }
+            LOG(INFO) << "Finished saving final optimized map";
+            return;
+        }
+
+        {
+            std::lock_guard<std::mutex> lk(map_builder_mutex);
+            // Save the map in a pbstream file
+            bool ok = map_builder.map_builder_->SerializeStateToFile(
+                true, filename_with_timestamp);
+            if (!ok) {
+                LOG(WARNING) << "Saving the map to pbstream failed.";
+            }
+        }
+        // Sleep for map_rate_sec duration, but check frequently for
+        // shutdown
+        while (b_continue_session) {
+            std::chrono::duration<double, std::milli> time_elapsed_msec =
+                std::chrono::high_resolution_clock::now() - start;
+            if ((time_elapsed_msec >= map_rate_sec) ||
+                (finished_processing_offline)) {
+                break;
+            }
+            if (map_rate_sec - time_elapsed_msec >=
+                check_for_shutdown_interval_usec) {
```

Comment:
@JohnN193 on shutdown timing

Reply:
what is this about?

Reply:
timing makes sense here. In orbslam Kat had setup the thread to loop at a faster rate in order to catch when a shutdown signal occurs, and this mirrors that implementation.